[Regression] libp2p host cannot be closed more than once (panic)

When a host is closed twice, go-eventbus triggers a panic:

```
[18:26:37] ~/g/s/g/i/ipfs-cluster (update/deps-asjnkwe|✚4…) $ go test 
--- FAIL: TestClustersStatusAllWithErrors (7.66s)
panic: closed an emitter more than once [recovered]
        panic: closed an emitter more than once

goroutine 271220 [running]:
testing.tRunner.func1(0xc000b0c000)
        /usr/lib64/go/1.12/src/testing/testing.go:830 +0x392
panic(0xf132e0, 0x1382950)
        /usr/lib64/go/1.12/src/runtime/panic.go:522 +0x1b5
github.com/libp2p/go-eventbus.(*emitter).Close(0xc00340fc80, 0xffffffffffffffff, 0xc000f87ea8)
        /home/hector/go/pkg/mod/github.com/libp2p/go-eventbus@v0.0.3/basic.go:40 +0x97
github.com/libp2p/go-libp2p/p2p/host/basic.(*BasicHost).Close(0xc003c2dc30, 0x0, 0xc0038d7af0)
        /home/hector/go/pkg/mod/github.com/libp2p/go-libp2p@v0.2.0/p2p/host/basic/basic_host.go:749 +0x3a
github.com/libp2p/go-libp2p/p2p/host/routed.(*RoutedHost).Close(0xc004048160, 0x0, 0x0)
        /home/hector/go/pkg/mod/github.com/libp2p/go-libp2p@v0.2.0/p2p/host/routed/routed.go:189 +0x33
github.com/ipfs/ipfs-cluster.shutdownClusters(0xc000b0c000, 0xc0046f3d10, 0x5, 0x5, 0xc0046f3cb0, 0x5, 0x5)
        /home/hector/go/src/github.com/ipfs/ipfs-cluster/ipfscluster_test.go:389 +0x120
github.com/ipfs/ipfs-cluster.TestClustersStatusAllWithErrors(0xc000b0c000)
        /home/hector/go/src/github.com/ipfs/ipfs-cluster/ipfscluster_test.go:769 +0x2bf
testing.tRunner(0xc000b0c000, 0x125b390)
        /usr/lib64/go/1.12/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
        /usr/lib64/go/1.12/src/testing/testing.go:916 +0x35a
exit status 2
FAIL    github.com/ipfs/ipfs-cluster    68.501s

```

This is new, did not happen before.

--- a/go.mod
+++ b/go.mod
@@ -21,11 +21,11 @@ require (
 	github.com/libp2p/go-libp2p-netutil v0.1.0
 	github.com/libp2p/go-libp2p-peerstore v0.1.1
 	github.com/libp2p/go-libp2p-secio v0.1.0
-	github.com/libp2p/go-libp2p-swarm v0.1.0
+	github.com/libp2p/go-libp2p-swarm v0.1.1
 	github.com/libp2p/go-libp2p-testing v0.0.4
 	github.com/libp2p/go-libp2p-transport-upgrader v0.1.1
 	github.com/libp2p/go-libp2p-yamux v0.2.1
-	github.com/libp2p/go-maddr-filter v0.0.4
+	github.com/libp2p/go-maddr-filter v0.0.5
 	github.com/libp2p/go-stream-muxer-multistream v0.2.0
 	github.com/libp2p/go-tcp-transport v0.1.0
 	github.com/libp2p/go-ws-transport v0.1.0

--- a/p2p/host/basic/basic_host.go
+++ b/p2p/host/basic/basic_host.go
@@ -747,6 +747,13 @@ func (h *BasicHost) AllAddrs() []ma.Multiaddr {
 
 // Close shuts down the Host's services (network, etc).
 func (h *BasicHost) Close() error {
+	// You're thinking of adding some teardown logic here, right? Well
+	// don't! Add any process teardown logic to the teardown function in the
+	// constructor.
+	//
+	// This:
+	// 1. May be called multiple times.
+	// 2. May _never_ be called if the host is stopped by the context.
 	return h.proc.Close()
 }
 

--- a/p2p/host/basic/basic_host.go
+++ b/p2p/host/basic/basic_host.go
@@ -146,6 +146,7 @@ func NewHost(ctx context.Context, net network.Network, opts *HostOpts) (*BasicHo
 		if h.cmgr != nil {
 			h.cmgr.Close()
 		}
+		_ = h.emitters.evtLocalProtocolsUpdated.Close()
 		return h.Network().Close()
 	})
 
@@ -746,7 +747,6 @@ func (h *BasicHost) AllAddrs() []ma.Multiaddr {
 
 // Close shuts down the Host's services (network, etc).
 func (h *BasicHost) Close() error {
-	_ = h.emitters.evtLocalProtocolsUpdated.Close()
 	return h.proc.Close()
 }
 

--- a/p2p/host/basic/basic_host_test.go
+++ b/p2p/host/basic/basic_host_test.go
@@ -23,6 +23,13 @@ import (
 	madns "github.com/multiformats/go-multiaddr-dns"
 )
 
+func TestHostDoubleClose(t *testing.T) {
+	ctx := context.Background()
+	h1 := New(swarmt.GenSwarm(t, ctx))
+	h1.Close()
+	h1.Close()
+}
+
 func TestHostSimple(t *testing.T) {
 	ctx := context.Background()
 	h1 := New(swarmt.GenSwarm(t, ctx))

--- a/p2p/host/basic/basic_host_test.go
+++ b/p2p/host/basic/basic_host_test.go
@@ -209,11 +209,19 @@ func TestHostProtoPreference(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	// XXX: This is racy now that we push protocol updates. If this tests
+	// fails, try allowing both protoOld and protoMinor.
 	assertWait(t, connectedOn, protoOld)
 
 	s2.Close()
 
 	s3, err := h2.NewStream(ctx, h1.ID(), protoMinor)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Force a lazy handshake as we may have received a protocol update by this point.
+	_, err = s3.Write([]byte("hello"))
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/p2p/host/basic/basic_host_test.go
+++ b/p2p/host/basic/basic_host_test.go
@@ -136,6 +136,8 @@ func TestHostAddrsFactory(t *testing.T) {
 }
 
 func getHostPair(ctx context.Context, t *testing.T) (host.Host, host.Host) {
+	t.Helper()
+
 	h1 := New(swarmt.GenSwarm(t, ctx))
 	h2 := New(swarmt.GenSwarm(t, ctx))
 
@@ -148,6 +150,7 @@ func getHostPair(ctx context.Context, t *testing.T) (host.Host, host.Host) {
 }
 
 func assertWait(t *testing.T, c chan protocol.ID, exp protocol.ID) {
+	t.Helper()
 	select {
 	case proto := <-c:
 		if proto != exp {

--- a/p2p/net/mock/mock_test.go
+++ b/p2p/net/mock/mock_test.go
@@ -11,6 +11,7 @@ import (
 	"time"
 
 	ci "github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/helpers"
 	"github.com/libp2p/go-libp2p-core/network"
 	"github.com/libp2p/go-libp2p-core/peer"
 	"github.com/libp2p/go-libp2p-core/protocol"
@@ -321,7 +322,7 @@ func TestStreams(t *testing.T) {
 func makePinger(st string, n int) func(network.Stream) {
 	return func(s network.Stream) {
 		go func() {
-			defer s.Close()
+			defer helpers.FullClose(s)
 
 			for i := 0; i < n; i++ {
 				b := make([]byte, 4+len(st))
@@ -342,7 +343,7 @@ func makePinger(st string, n int) func(network.Stream) {
 func makePonger(st string) func(network.Stream) {
 	return func(s network.Stream) {
 		go func() {
-			defer s.Close()
+			defer helpers.FullClose(s)
 
 			for {
 				b := make([]byte, 4+len(st))

